### PR TITLE
Call 'onXXXChange' event on 'onOptionChange'

### PR DESCRIPTION
**Pull request by LazyLahtak**

### CODE REVIEW

File: src/core/__tests__/props-updating.test.tsx [hunk 1]
```diff
@@ -619,3 +623,123 @@ describe("mutation detection", () => {
         expect(Widget.option.mock.calls[0][1]).toEqual(value);
     };
 });
+
+describe("onXXXChange", () => {
+    it("should call on component changes option", () => {
```
Comment (IlyaKhD) [inline src/core/__tests__/props-updating.test.tsx:628]:
```suggestion
    it("is called on component changes option", () => {
```
Same for other tests

File: src/core/options-manager.ts [hunk 2]
```diff
@@ -124,6 +126,56 @@ class OptionsManager {
...
+            return option;
+        }
+
+        return null;
+    }
+
+    private _callOptionChangeHandler(optionName: string, optionValue: any) {
+        const parts = optionName.split(".");
+        let eventName = parts[parts.length - 1];
+        if (eventName.substr(0, 2) !== "on") {
```
Comment (IlyaKhD) [inline src/core/options-manager.ts:157]:
```suggestion
        if (eventName.startsWith("on")) {
            return
        }
```

File: src/core/options-manager.ts [hunk 3]
```diff
@@ -124,6 +126,56 @@ class OptionsManager {
...
+            }
+            return option;
+        }
+
+        return null;
+    }
+
+    private _callOptionChangeHandler(optionName: string, optionValue: any) {
+        const parts = optionName.split(".");
+        let eventName = parts[parts.length - 1];
```
Comment (IlyaKhD) [inline src/core/options-manager.ts:156]:
```suggestion
        const propName = parts[parts.length - 1];
```

File: src/core/options-manager.ts [hunk 4]
```diff
@@ -124,6 +126,56 @@ class OptionsManager {
...
+        if (eventName.substr(0, 2) !== "on") {
+            eventName = `on${capitalizeFirstLetter(eventName)}Change`;
```
Comment (IlyaKhD) [inline src/core/options-manager.ts:158]:
```suggestion
            const eventName = `on${capitalizeFirstLetter(propName)}Change`;
```

File: src/core/options-manager.ts [hunk 5]
```diff
@@ -124,6 +126,56 @@ class OptionsManager {
...
+                { ...this._currentConfig.configs, ...this._currentConfig.configCollections }
+            );
+
+            if (changeEvent) {
+                if (typeof changeEvent === "function") {
+                    changeEvent(optionValue);
+                } else {
+                    throw new Error(
+                        `Invalid value for "${eventName}" property.
+                        The "${eventName}" must be a function.`
```
Comment (IlyaKhD) [inline src/core/options-manager.ts:172]: @RomanTsukanov, could you check this message?

File: src/core/options-manager.ts [hunk 6]
```diff
@@ -124,6 +126,56 @@ class OptionsManager {
...
+        if (eventName.substr(0, 2) !== "on") {
+            eventName = `on${capitalizeFirstLetter(eventName)}Change`;
+            parts[parts.length - 1] = eventName;
+            const changeEvent = this._getOptionFromConfig(
+                parts,
+                this._currentConfig.options,
+                { ...this._currentConfig.configs, ...this._currentConfig.configCollections }
+            );
+
+            if (changeEvent) {
```
Comment (IlyaKhD) [inline src/core/options-manager.ts:166]:
```suggestion
            if (!changeEvent) {
                return;
            }
```

File: src/core/options-manager.ts [hunk 7]
```diff
@@ -124,6 +126,56 @@ class OptionsManager {
...
+                if (typeof changeEvent === "function") {
+                    changeEvent(optionValue);
+                } else {
+                    throw new Error(
+                        `Invalid value for "${eventName}" property.
+                        The "${eventName}" must be a function.`
+                    );
+                }
```
Comment (IlyaKhD) [inline src/core/options-manager.ts:174]:
```suggestion
                if (typeof changeEvent !== "function") {
                    throw new Error(
                        `Invalid value for "${eventName}" property.
                        The "${eventName}" must be a function.`
                    );
                }
                
               changeEvent(optionValue);
```

File: src/core/options-manager.ts [hunk 8]
```diff
@@ -124,6 +126,28 @@ class OptionsManager {
...
+            parts[parts.length - 1] = eventName;
+            const changeEvent = findValue(this._currentConfig, parts);
+
+            if (!changeEvent) {
+                return;
+            }
+
+            if (typeof changeEvent.value !== "function") {
+                throw new Error(
+                    `Invalid value for "${eventName}" property.
```
Comment (RomanTsukanov) [inline src/core/options-manager.ts:143]:
```suggestion
                    `Invalid value for the ${eventName} property.
```

File: src/core/options-manager.ts [hunk 9]
```diff
@@ -124,6 +126,28 @@ class OptionsManager {
...
+                    The "${eventName}" must be a function.`
```
Comment (RomanTsukanov) [inline src/core/options-manager.ts:144]:
```suggestion
                    ${eventName} must be a function.`
```

**Review by ovchinnikov — APPROVED** [review]

File: src/core/__tests__/props-updating.test.tsx [hunk 10]
```diff
@@ -619,3 +629,123 @@ describe("mutation detection", () => {
...
+        component.setProps(sampleProps);
+
+        expect(onPropChange).not.toBeCalled();
+
+        fireOptionChange("text", "2");
+        expect(onPropChange).toHaveBeenCalledTimes(1);
+        expect(onPropChange).toBeCalledWith("2");
+    });
+
+    it("should call on component changes complex option", () => {
```
Comment (IlyaKhD) [inline src/core/__tests__/props-updating.test.tsx:654]:
```suggestion
    it("is called on component changes complex option", () => {
```

File: src/core/__tests__/props-updating.test.tsx [hunk 11]
```diff
@@ -619,3 +629,123 @@ describe("mutation detection", () => {
...
+            />
+        );
+        expect(onPropChange).not.toBeCalled();
+
+        fireOptionChange("complexOption.text", "1");
+        expect(onPropChange).toHaveBeenCalledTimes(1);
+        expect(onPropChange).toBeCalledWith("1");
+    });
+
+    it("should call on component changes array option", () => {
```
Comment (IlyaKhD) [inline src/core/__tests__/props-updating.test.tsx:668]:
```suggestion
    it("is called on component changes array option", () => {
```

File: src/core/__tests__/props-updating.test.tsx [hunk 12]
```diff
@@ -619,3 +629,123 @@ describe("mutation detection", () => {
...
+        expect(onSecondPropChange).not.toBeCalled();
+        expect(onFirstPropChange).toHaveBeenCalledWith("1");
+
+        fireOptionChange("arrayOption[1].text", "2");
+        expect(onFirstPropChange).toHaveBeenCalledTimes(1);
+        expect(onSecondPropChange).toHaveBeenCalledTimes(1);
+        expect(onSecondPropChange).toHaveBeenCalledWith("2");
+    });
+
+    it("should call on nested option changed", () => {
```
Comment (IlyaKhD) [inline src/core/__tests__/props-updating.test.tsx:694]:
```suggestion
    it("is called on nested option changed", () => {
```

File: src/core/__tests__/props-updating.test.tsx [hunk 13]
```diff
@@ -619,3 +629,123 @@ describe("mutation detection", () => {
...
+        fireOptionChange("nestedOption.value", "3");
+        expect(onNestedPropChange).toHaveBeenCalledTimes(1);
+        expect(onNestedPropChange).toBeCalledWith("3");
+
+        fireOptionChange("items[1].nestedOption.value", "4");
+        expect(onSubNestedPropChange).toHaveBeenCalledTimes(1);
+        expect(onSubNestedPropChange).toBeCalledWith("4");
+    });
+
+    it("should throw an error if handler is not a function", () => {
```
Comment (IlyaKhD) [inline src/core/__tests__/props-updating.test.tsx:741]:
```suggestion
    it("throws an error if handler is not a function", () => {
```

File: src/core/options-manager.ts [hunk 14]
```diff
@@ -124,6 +126,28 @@ class OptionsManager {
         }
     }
 
+    private _callOptionChangeHandler(optionName: string, optionValue: any) {
+        const parts = optionName.split(".");
+        const propName = parts[parts.length - 1];
+        if (propName.substr(0, 2) !== "on") {
```
Comment (IlyaKhD) [inline src/core/options-manager.ts:132]:
```suggestion
        if (eventName.startsWith("on")) {
            return;
        }
```

**Review by ovchinnikov — APPROVED** [review]

**Pull request merged**